### PR TITLE
Added transaction trace (#3076)

### LINKED ISSUE #3076: How can we trace BeginTransaction(), RollbackTransaction() and CommitTransaction()?
### Discussed in https://github.com/linq2db/linq2db/discussions/3073

<div type='discussions-op-text'>

<sup>Originally posted by **ww9** June 24, 2021</sup>
Hi,
I've been trying to log SQL queries like this and it works but it doesn't tell me about transactions which are important to my log.

# Environment details
linq2db version: Release 4.0.0-preview.1
Database Server: MariaDB 10.4
Database Provider: MySql.Data 8.0.25
Operating system: Windows 10 x64
.NET Framework: .NET 5.0

# Trace setup
```csharp
LinqToDB.Data.DataConnection.TraceSwitch.Level = TraceLevel.Verbose;
LinqToDB.Data.DataConnection.OnTrace = (TraceInfo info) =>
{
   System.Diagnostics.Debug.WriteLine("\n\n" + info.Operation.ToString() + "." 
      + info.TraceInfoStep + " @@@@@@@@@@@@@@@@\n" + info.SqlText);
};
LinqToDB.Data.DataConnection.TurnTraceSwitchOn();
```

# Application code
```csharp
db.BeginTransaction();
db.UserAgent.Value(u => u.Hash, "hash").Value(u => u.AgentString, "agent").Insert();
db.RollbackTransaction();
```

# Output (missing any information about transactions):

```
ExecuteNonQuery.BeforeExecute @@@@@@@@@@@@@@@@
-- Default MySql.Official MySql
DECLARE @Hash VarChar(4) -- String
SET     @Hash = 'hash'
DECLARE @AgentString VarChar(5) -- String
SET     @AgentString = 'agent'

INSERT INTO `app`.`user_agent`
(
	`Hash`,
	`AgentString`
)
VALUES
(
	@Hash,
	@AgentString
)

ExecuteNonQuery.AfterExecute @@@@@@@@@@@@@@@@
-- Default MySql.Official MySql
DECLARE @Hash VarChar(4) -- String
SET     @Hash = 'hash'
DECLARE @AgentString VarChar(5) -- String
SET     @AgentString = 'agent'

INSERT INTO `app`.`user_agent`
(
	`Hash`,
	`AgentString`
)
VALUES
(
	@Hash,
	@AgentString
)

DisposeQuery.Completed @@@@@@@@@@@@@@@@
-- Default MySql.Official MySql
DECLARE @Hash VarChar(4) -- String
SET     @Hash = 'hash'
DECLARE @AgentString VarChar(5) -- String
SET     @AgentString = 'agent'

INSERT INTO `app`.`user_agent`
(
	`Hash`,
	`AgentString`
)
VALUES
(
	@Hash,
	@AgentString
)
```

How can I trace transaction start, rollback and commit?

</div>

### PR DESCRIPTION
**Pull request by SebastianBienert**
Fix #3076

### CODE REVIEW

**Comment by MaceWindu** [comment]
Thanks, will review it a bit later

**Review by MaceWindu — APPROVED** [review]

File: Source/LinqToDB/Data/DataConnection.Async.cs [hunk 1]
```diff
@@ -26,18 +26,28 @@ public virtual async Task<DataConnectionTransaction> BeginTransactionAsync(Cance
 			//
 			if (TransactionAsync != null) await TransactionAsync.DisposeAsync().ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);
 
-			// Create new transaction object.
-			//
-			TransactionAsync = await _connection!.BeginTransactionAsync(cancellationToken).ConfigureAwait(Common.Configuration.ContinueOnCapturedContext);
+			var dataConnectionTransaction = await TraceActionAsync(
```
Comment (viceroypenguin) [inline Source/LinqToDB/Data/DataConnection.Async.cs:29]: Is there a reason this `BeginTransactionAsync()` doesn't pass through to the other `BeginTransactionAsync()` with appropriate parameter set?

Comment (MaceWindu) [reply Source/LinqToDB/Data/DataConnection.Async.cs:29]: You mean with default IsolationLevel? If yes, then there is no such thing as default isolation level

**Review by viceroypenguin — APPROVED** [review]

File: Source/LinqToDB/Data/DataConnection.Async.cs (hunk 1 above)
Comment (viceroypenguin) [reply Source/LinqToDB/Data/DataConnection.Async.cs:29]: 👍🏼

File: Source/LinqToDB/Data/DataConnection.Async.cs [hunk 2]
```diff
@@ -26,18 +26,28 @@ public virtual async Task<DataConnectionTransaction> BeginTransactionAsync(Cance
...
-			if (_command != null)
-				_command.Transaction = Transaction;
+					// If the active command exists.
+					//
+					if (dataConnection._command != null)
+						dataConnection._command.Transaction = dataConnection.Transaction;
 
-			return new DataConnectionTransaction(this);
+					return new DataConnectionTransaction(dataConnection);
+				}, (object?)null, cancellationToken)
```
Comment (viceroypenguin) [inline Source/LinqToDB/Data/DataConnection.Async.cs:47]: Slightly preferred according to newer style standards, but no real preference either way.

```suggestion
				}, default(object?), cancellationToken)
```

**Review by sdanyliv — APPROVED** [review]

**Comment by MaceWindu** [comment]
postmortem on baselines: except explicit transactions called in tests there are also logged transactions for eager load queries

**Pull request merged**